keps/sig-node/4569-cgroup-v1-deprecation/README.md

```diff
@@ -0,0 +1,685 @@
+
+
+#### Code Modifications for Warning Log Implementation
+
+To incorporate these warning logs, the kubelet's code will be adjusted in locations where the IsCgroup2UnifiedMode() function is invoked. e.g.
+
+Original Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+		if !isCgroup2UnifiedMode() {
+			swapConfigurationHelper.ConfigureNoSwap(lcr)
+			return
+		}
+	}
+```
+Revised Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+    if !isCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) && !SupressCgroupV1Warning() {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      eventRecorder.Event(pod, v1.EventTypeWarning, "DeprecatedCgroupV1", fmt.Sprint("cgroup v1 support has been deprecated"))
+      swapConfigurationHelper.ConfigureNoSwap(lcr)
+      return
+    }
+}
+```
+
+Furthermore, code segments that default to cgroup v1 logic will be inverted to assume cgroup v2 as the default. This shift underscores the transition to cgroup v2 and ensures warning logs are consistently generated for cgroup v1 usage. e.g.
+
+Original Code Snippet:
+```golang
+	memLimitFile := "memory.limit_in_bytes"
+	if libcontainercgroups.IsCgroup2UnifiedMode() {
+		memLimitFile = "memory.max"
+	}
+```
+Revised Code Snippet:
+```golang
+	memLimitFile = "memory.max"
+	if libcontainercgroups.!IsCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) && !SupressCgroupV1Warning() {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      eventRecorder.Event(pod, v1.EventTypeWarning, "DeprecatedCgroupV1", fmt.Sprint("cgroup v1 support has been deprecated"))
+      memLimitFile := "memory.limit_in_bytes"
+	}
+```
+
+### Feature Flag and Kubelet Option
+
+During the `alpha` stage of this feature's lifecycle, it's anticipated that users will require time to plan and execute their migrations to cgroup v2. Factors influencing the migration timeline include, but are not limited to, selecting compatible kernels and operating systems, planning for potential downtimes, and coordinating changes to minimize disruption during peak operational periods.
+
+To accommodate this, we propose introducing a kubelet flag, `--suppress-cgroupv1-deprecation-warning`, which is not set by default. This flag provides users the flexibility to suppress repetitive deprecation warning logs and events related to cgroup v1 while they are in the process of planning and executing their migration to cgroup v2. The introduction of this flag is intended to reduce log spam and operational noise, acknowledging the non-trivial nature of the migration for many users.
```

Comment:
I'm not convinced that we want to go this route..

Do we really want to encourage people to turn off a deprecation? We are warning them because we plan to make this a fail case. If we are worried about log spam, we could always maybe it a V(3) or above log. 

Maybe its something we can leave open for discussion.

Reply:
Instead of turning off completely, changing the log level sounds good.

Reply:
and we don't have to add any extra kubelet arguments to suppress the logs either.

Reply:
Something that doesn't change its value during while the kubelet is up and running, there is no need to bother the user over and over about it, right? We come across like nagging telemarketers if we do that. ;)

Reply:
Yea I think logging once on startup and don't allow one to turn it off is a good idea!

Reply:
Agreed!

Reply:
what Kevin says 👍 , we can not add a temporary flag that we'll need to remove later because it will be a noop once cgroupsv1 was completely deprecated